### ISSUE TITLE
Feature Request: implement Moderation system

### ISSUE THREAD

**Issue by ingoj** [friendica/friendica#8724]
I think it would be good to have a moderation system similar to Mastodon where you can report problematic content, an admin will be notified and decide what to do with that reported post: 
- warn the user 
- silence the user (content won't be listed in public timelines anymore)
- suspend the user and delete the content (i.e. deleting user and its content)

Currently the admin can only delete the user or block the user in Friendica, but there is no way that other users can report unwanted content to the admins, except directly reaching out to the admin. The admin on the other hand has no real option to notify an user about reported posts: he/she cannot delete the reported content easily nor sending a message is easy without first making a contact with that user and then sending a private message. 

Often I just delete users without prior notification, which I find a little bit too harsh, because I have simply not the time to manually interact with that user with the given options at the moment in Friendica. 

It would also be good to expand bin/console by those features to: 
- warn a user
- hide a specific post
- delete a specific post
- suspend a user (maybe for a given time frame)
- silence a user
- etc.

**Comment by copiis** [comment]
I think that's a very useful function

**Comment by AndyHee** [comment]
Yep, that would useful.

**Comment by Singh-in** [comment]
If an account is reported on Mastodon and the account is located on a remote server, then this server can be notified. 
Does this message arrive at all? I tried this on 30.12.2021 and until today no message has arrived at the admin.

In light of the ongoing discussion within the EU, the problem needs to be addressed very quickly. Otherwise, the Friendica admin will probably get into bigger legal problems.

**Comment by MrPetovan** [comment]
Currently remote reports from Mastodon should be forwarded to Friendica nodes but we don't handle them yet as we don't even have any local reporting feature yet.

**Comment by Singh-in** [comment]
This case shows that Friendica is not only used by friendly people. Therefore, the basic features and requirements should be described and implemented. 

https://framagit.org/tom79/fedilab/-/issues/498
https://f-droid.org/en/2019/07/16/statement.html

**Comment by Quix0r** [comment]
Further ideas (needs implementation in Mastodon, Diaspora*, etc.):

- Federated voting on those reports, to avoid trolls/abuse
- Federated light commenting
- Openness to the reported user/post by a notification
- Not all administrators/moderators speak/read all languages (me: DE, EN) and wish to use external automated translation (e.g. deepl)

**Comment by informapirata** [comment]
It would be a key feature. Unfortunately, user management is friendica's real weakness.

I do not expect advanced features like those of Mastodon, but at least as an administrator I would like to read the reports made by other instances (the risk is that if I do not notice that I have a problematic user, I could be blocked as an instance) and I would like the possibility to warn a problematic user. And my users have a hard time reporting other users' problematic behaviors.

**Comment by MrPetovan** [comment]
I spent way too long finding it back again today, so pasting the link to the Moderation feature CryptPad I opened a couple months ago: https://office.trom.tf/pad/#/2/pad/edit/1ensaAqGa8BSt3yPNBftb+mr/

**Comment by annando** [comment]
I guess having the pure basics of the pad would be a big step forward.

**Comment by MrPetovan** [comment]
Yes, I was about to get on it but I couldn't find the document again.

**Comment by FediVideos** [comment]
Is there any chance of a report button being added?

I'd love to promote Friendica as a place for new people to join, but absence of a reporting system is a deal-breaker, at least for me.

**Comment by MrPetovan** [comment]
Yes, it's intended as part of this feature.

**Comment by AlfredSK** [comment]
From the perspective of an admin/moderator:

Reports for posts/comments are easy to handle. You receive the report. You have a link to a specific item. You can have a look at that item and make your decision. Quick and easy.
Reports for users are really hard. Often you just receive a report for a user with a description like "hater" or "nazi". When you follow the link to the user's profile you are confronted with hundreds or even thousands of posts/comments about linux or other tech stuff. You have to scroll through all that content to (maybe) find the reason for the report. Time consuming and hard.

**Comment by MrPetovan** [comment]
Technically, any number of posts can be attached to a moderation report, so there's little reason for an admin/moderator to spend too much time on a succinct report. We are planning a public comment field where moderators can ask the reporter for more information but this would only hold for local node reporters.

**Comment by FediVideos** [comment]
Would it help if there was a simple message for people doing reports directly from a profile asking people to report the problematic posts if they can?

And, if they report via a specific post this message wouldn't appear?

I think some people are just unaware how time-consuming it can be to find offending posts if a report is via a profile.

Having said that, it's useful to have a profile reporting system as some accounts simply boost offensive content and set up new accounts to boost if one gets shut down. This allows them to build a following of nasty people without being directly held responsible.

**Comment by MrPetovan** [comment]
I’m working towards a moderation report creation wizard, including a step where posts by the reported contact can be attached.

**Comment by scarlet-tobar** [comment]
I'm interested in making a report/moderation addon (not promising anything), but I'm not sure where to start. What do you suggest to investigate about the source code and similar stuff? Thanks

**Comment by MrPetovan** [comment]
Thank you for your interest! I suggest you start with a small feature, like silencing a user from the user list, and figure out how silencing is done from the command line and how blocking is done from the user list and copy the feature for silencing.

Let's go on a magical journey:
1. Silencing is done from the `globalcommunitysilence` console command.
2. This command is routed to the `Friendica\Console\GlobalCommunitySilence` class in the `src/Core/Console.php` file.
3. This class mainly retrieves a `contact` record based on a URL and updates its `hidden` field.
4. Now follow the code path from the path `/moderation/users`:
1. We first use the `static/routes.config.php` router file to figure out what module class will be invoked, in this case `Friendica\Module\Moderation\Users\Index`.
4. A module can respond differently to GET requests (in `content()`) or POST requests (in `post()`). The list display is created in `content()`.
5. A list of users is retrieved with `Friendica\Model\User::getList` and passed as a variable to the `moderation/users/index.tpl` Smarty template. This file can either be in `view/templates/moderation/users/index.tpl` or overridden `view/themes/<current theme>/templates/moderation/users/index.tpl`
6. Let's consider `view/theme/frio/templates/moderation/users/index.tpl`. The variables are invoked using `$variable` and can be simply interpolated in the resulting HTML with `{{$variable}}`.
7. It's also possible to loop using `{{foreach}}` and use conditions with `{{if}}`.
8. Notice the `{{$baseurl}}/moderation/users/block/{{$u.uid}}` link to block a single user.
9. Back to the router file, we figure out that it uses the `/users[/{action}/{uid}]` pattern with the optional `action` and `uid` parameters which value will be available in the `Friendica\Module\Moderation\Users\Index` class in the property `parameters`.
10. Following the switch on the `$action` variable, we figure out that we're calling `Friendica\Model\User::block` using the `uid` value we got from the user record in the template loop.
11. Since we need to update the contact record associated with a local user, we can use the owner abstraction (provided by the `owner-view` view) that combines both, using `Friendica\Model\User::getOwnerByUid` to retrieve the combined record.
12. Then it's a matter of updating the `hidden` field of the `contact` row referenced by the `owner.id` value.

Displaying the current silence status of a local user would also need to use the owner abstraction and is left as an exercise to the reader.

**Comment by scarlet-tobar** [comment]
Thanks! I'll see if I can start after my uni evaluations.

**Comment by scarlet-tobar** [comment]
I made a little SilenceServer for console to start but it's pretty basic and I'm not sure it works (not too much experience with PHP).

It's here https://github.com/scarlet-tobar/friendicaF/blob/task/moderation-reports/src/Console/ServerSilence.php

I'll try to make more things in the next days.

Thanks for your work.

**Comment by MrPetovan** [comment]
A cursory look tells me it could work, but any use of it would wipe all the single contact silencing decisions made beforehand.

Instead, the `hidden` field should be set at the `gserver` record level, and checked when building the global timeline.

**Comment by scarlet-tobar** [comment]
Gserver has a hidden field? I think i did not see it.

**Comment by MrPetovan** [comment]
Probably not, but it can be added by updating the database structure schema in the `static/dbstructure.config.php` file and then running the `bin/console dbstructure dump > database.sql` to update the SQL schema and the database documentation files that should be part of the pull request.

This workflow requires a working Friendica install, let me know if you'd like help for this.

**Comment by scarlet-tobar** [comment]
Thanks, how do i make an development install of friendica? (on localhost or similar?)

**Comment by MrPetovan** [comment]
Please take a look first at the existing [installation documentation](https://github.com/friendica/friendica/blob/develop/doc/Install.md), I'm available to answer specific questions about it.

**Comment by tobiasd** [comment]
You can also run a development node using [Vagrant](https://github.com/friendica/friendica/blob/develop/doc/Vagrant.md)

**Comment by FediVideos** [comment]
Just got told the moderation system has been implemented, is this correct? Is this issue done?

**Comment by scarlet-tobar** [comment]
> Just got told the moderation system has been implemented, is this correct? Is this issue done?

Partially. Not all features above in the issue are available.

**Comment by FediVideos** [comment]
Can I just check, has the moderation system been implemented? 🙂

(Asking as there was a discussion about this on FediTips.)

**Comment by MrPetovan** [comment]
I don't think any work has been added to this issue. I unassigned myself to dispel any confusion. Hopefully if work is done towards this feature, pull requests will reference this issue so that you can see the progress.

**Comment by annando** [comment]
@FediVideos we are currently running with very low development capacity. Some additional help would be needed to implement something like this.

**Comment by FediVideos** [comment]
Okay, thanks for the updates on the situation 🙏 Sorry to hear about the low capacity, do you want people to raise awareness about Friendica to attract more contributors?

**Comment by annando** [comment]
Good question. I'm bad at marketing.

**Comment by hoergen** [comment]
> Okay, thanks for the updates on the situation 🙏 Sorry to hear about the low capacity, do you want people to raise awareness about Friendica to attract more contributors?

yes, that would be nice. If you have some ideas and see some ways to increase the visibility and attract more developers ❤️
